File: ios/nodes/GainNode/GainNode.m
#import <GainNode.h>
#import "AudioContext.h"
#import "Constants.h"

@implementation GainNode

- (instancetype)initWithContext:(AudioContext *)context {
    if (self = [super initWithContext:context]) {
        _gainParam = [[AudioParam alloc] initWithContext:context value:1 minValue:-[Constants maxGain] maxValue:[Constants maxGain]];
        self.numberOfInputs = 1;
        self.numberOfOutputs = 1;
    }

    return self;
}

- (void)cleanup {
    _gainParam = nil;
}

- (void)processWithParameters:(PlaybackParameters *)parameters {
    float currentTime = [self.context getCurrentTime];
    parameters.leftGain *= [_gainParam getValueAtTime:currentTime];
    parameters.rightGain *= [_gainParam getValueAtTime:currentTime];

    [super processWithParameters:parameters];
}

@end

File: android/src/main/java/com/swmansion/audioapi/nodes/GainNode.kt
package com.swmansion.audioapi.nodes

import android.os.Build
import androidx.annotation.RequiresApi
import com.swmansion.audioapi.context.BaseAudioContext
import com.swmansion.audioapi.nodes.audionode.AudioNode
import com.swmansion.audioapi.parameters.AudioParam
import com.swmansion.audioapi.parameters.PlaybackParameters
import com.swmansion.audioapi.utils.Constants

class GainNode(
  context: BaseAudioContext,
) : AudioNode(context) {
  override val numberOfInputs: Int = 1
  override val numberOfOutputs: Int = 1
  override var channelCount: Int = 2

  private val gain: AudioParam = AudioParam(context, 1.0, Constants.MAX_GAIN, -Constants.MAX_GAIN)
    get() = field

  @RequiresApi(Build.VERSION_CODES.N)
  override fun process(playbackParameters: PlaybackParameters) {
    mixBuffers(playbackParameters)

    val gain = this.gain.getValueAtTime(context.getCurrentTime())
    playbackParameters.rightPan *= gain
    playbackParameters.leftPan *= gain

    super.process(playbackParameters)
  }
}
